### PR TITLE
:sparkles: Middleware addition

### PR DESCRIPTION
**Pull request by Endercheif**
### Changes

-   `adds`: 
    - `on_channel_delete`
    - `on_channel_pins_update`
    - `on_channel_update`
    - `on_guild_delete`
    - `on_guild_update`
    - `on_message_reaction_add`
    - `on_message_reaction_remove`
    - `on_message_reaction_remove_all`
    - `on_message_reaction_remove_emoji`
    - `on_thread_create`
    - `on_thread_delete`
    - `on_thread_list_sync`
    - `on_thread_member_update`
    - `on_thread_members_update`
    - `on_thread_update`

 ### Check off the following

-   [x] I have tested my changes with the current requirements
-   [x] My Code follows the pep8 code style.

### CODE REVIEW

**Comment by Endercheif** [comment]
dont review yet. im just fixing that last commit

**Comment by Endercheif** [comment]
fixed

**Comment by Endercheif** [comment]
commit was wacky. I'm fixing it

**Comment by Arthurdw** [comment]
https://app.codacy.com/gh/Pincer-org/Pincer/pullRequest?prid=8299424

**Comment by Endercheif** [comment]
> 
> 
> https://app.codacy.com/gh/Pincer-org/Pincer/pullRequest?prid=8299424

@Arthurdw 👍

File: pincer/middleware/channel_delete.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    payload : :class:`pincer.core.dispatch.GatewayDispatch`
+        The data received from the channel delete event.
+
+    Returns
+    -------
+    Tuple[:class:`str`, `~pincer.objects.guild.channel.Channel`]
+        ``on_channel_delete`` and a ``Channel``
+    """
+    return "on_channel_delete", [
```
Comment (Arthurdw) [inline pincer/middleware/channel_delete.py:29]: If client has the channel guild cached, remove the channel from that guild.

File: pincer/middleware/channel_update.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        The data received from the channel update event.
+
+
+    Returns
+    -------
+    Tuple[:class:`str`, List[:class:`~pincer.objects.channel.channel.Channel`]]
+        ``on_channel_update`` and a ``Channel``
+    """
+
+    return "on_channel_update", [
```
Comment (Arthurdw) [inline pincer/middleware/channel_update.py:31]: If client has the channel guild cached, update channel.

File: pincer/middleware/guild_delete.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    payload : :class:`GatewayDispatch`
+        The data received from the guild delete event.
+
+
+    Returns
+    -------
+    Tuple[:class:`str`, List[:class:`~pincer.objects.guild.guild.UnavailableGuild`]]
+        ``on_guild_delete`` and an ``UnavailableGuild``
+    """
+
```
Comment (Arthurdw) [inline pincer/middleware/guild_delete.py:30]: Delete guild from client cache

File: pincer/middleware/guild_update.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    payload : :class:`GatewayDispatch`
+        The data received from the guild update event.
+
+
+    Returns
+    -------
+    Tuple[:class:`str`, List[:class:`~pincer.objects.guild.guild.Guild`]]
+        ``on_guild_Update`` and an ``Guild``
+    """
+
```
Comment (Arthurdw) [inline pincer/middleware/guild_update.py:31]: Update guild in client cache

File: pincer/middleware/guild_update.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    channel_list = payload.data.pop("channels", [])
+
+    channels: List[Channel] = [
+        Channel.from_dict(construct_client_dict(self, channel))
+        for channel in channel_list
+    ]
```
Comment (Arthurdw) [inline pincer/middleware/guild_update.py:37]: Use mapping

File: pincer/middleware/thread_list_sync.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    Returns
+    -------
+    Tuple[:class:`str`, List[:class:`~pincer.objects.guild.events.thread.ThreadListSyncEvent`]]
+        ``on_thread_list_sync`` and an ``ThreadListSyncEvent``
+    """
+
+    threads: List[Channel] = [
+        Channel.from_dict(construct_client_dict(self, thread))
+        for thread in payload.data.pop("threads")
+    ]
```
Comment (Arthurdw) [inline pincer/middleware/thread_list_sync.py:39]: Use mapping

File: pincer/middleware/thread_list_sync.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    members: List[ThreadMember] = [
+        ThreadMember.from_dict(construct_client_dict(self, member))
+        for member in payload.data.pop("members")
+    ]
```
Comment (Arthurdw) [inline pincer/middleware/thread_list_sync.py:44]: Use mapping

File: pincer/middleware/thread_members_update.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    added_members: List[ThreadMember] = [
+        ThreadMember.from_dict(construct_client_dict(
+            self,
+            {
+                "join_timestamp": Timestamp(added_member.pop("join_timestamp")),
+                **added_member
+            }
+        ))
+        for added_member in payload.data.pop("added_members")
+    ]
```
Comment (Arthurdw) [inline pincer/middleware/thread_members_update.py:43]: Use mapping

File: pincer/objects/events/channel.py [hunk 9]
```diff
@@ -28,3 +29,9 @@ class ChannelPinsUpdateEvent(APIObject):
 
     guild_id: APINullable[Snowflake] = MISSING
     last_pin_timestamp: APINullable[Timestamp] = MISSING
+
+
+    def __post_init__(self):
+        self.last_pin_timestamp = convert(
+            self.last_pin_timestamp, Timestamp.parse, Timestamp
+    )
```
Comment (Arthurdw) [inline pincer/objects/events/channel.py:37]: Post init's arn't required on `APIObject`'s as these should get casted automatically.

File: pincer/objects/message/emoji.py [hunk 10]
```diff
@@ -45,8 +45,8 @@ class Emoji(APIObject):
         user that created this emoji
     """
 
-    id: Optional[Snowflake]
     name: Optional[str]
+    id: Optional[Snowflake] = None
```
Comment (Arthurdw) [inline pincer/objects/message/emoji.py:49]: Use `APINullable` & `MISSING` for optional parameters.

**Review by Arthurdw — CHANGES_REQUESTED** [review]

**Comment by Endercheif** [comment]
@Arthurdw can this be merged?

**Review by Arthurdw — APPROVED** [review]

**Pull request merged**